src/enzyme_ad/jax/gpu.cc

```diff
@@ -106,10 +111,65 @@ XLA_FFI_Error *execute(XLA_FFI_CallFrame *call_frame) {
   return nullptr;
 }
 
+XLA_FFI_Error *execute_with_error(XLA_FFI_CallFrame *call_frame) {
+  // If passed a call frame with the metadata extension, just return the
+  // metadata.
+  if (call_frame->extension_start != nullptr &&
+      call_frame->extension_start->type == XLA_FFI_Extension_Metadata) {
+    auto extension = reinterpret_cast<XLA_FFI_Metadata_Extension *>(
+        call_frame->extension_start);
+    extension->metadata->api_version = XLA_FFI_Api_Version{
+        XLA_FFI_Api_Version_STRUCT_SIZE,
+        /*extension_start=*/nullptr,
+        XLA_FFI_API_MAJOR,
+        XLA_FFI_API_MINOR,
+    };
+    return nullptr;
+  }
+
+  auto *bspan =
+      reinterpret_cast<XLA_FFI_ByteSpan *>(call_frame->attrs.attrs[0]);
+  auto cinfo = (CallInfoWithError *)bspan->ptr;
+
+  auto internal_api = call_frame->api->internal_api;
+  auto ctx = call_frame->ctx;
+  void *stream =
+      ((stream_executor::Stream *)internal_api->XLA_FFI_INTERNAL_Stream_Get(
+           ctx))
+          ->platform_specific_handle()
+          .stream;
+
+  size_t numargs = call_frame->args.size;
+  void *ptrs[numargs];
+  for (size_t i = 0; i < numargs; i++) {
+    ptrs[i] =
+        &reinterpret_cast<XLA_FFI_Buffer *>(call_frame->args.args[i])->data;
+  }
+
+  auto *execution_state = reinterpret_cast<xla::ffi::ExecutionState *>(
+      internal_api->XLA_FFI_INTERNAL_ExecutionState_Get(ctx));
```

Comment:
maybe we could push all the common code into a common static function (maybe with a template so we could do 
```
if constexpr (with_errors) {
   error code
} else {
   original code
}
```